### PR TITLE
[#149806558] Compose metric scraping and reporting to Datadog

### PR DESCRIPTION
**Pull request by 46bit**
## What

We have implemented a scraper for the metrics of our Compose cluster (as used for production.) This uses the Scrapey library for Python to perform the HTTPS requests, extract the relevant elements and extract the RAM and Disk metrics for each host.

An app to be deployed on the PaaS runs the scraper and metric reporting periodically. Metrics are logged to Datadog:

* Metrics of the `in_use` (percentage used), `used` (GB used) and `total` (GB total) for RAM and Disk for every Compose host.
* Service check that the service is alive, and any failed scraping attempts.

For more information see the commit messages and the new `README.md`.

## How to review

* Probably best to review in conjunction with [paas-cf PR1036](https://github.com/alphagov/paas-cf/pull/1036) that deploys this app and create Datadog monitors to alert on this data. 

## Who can review

Not @46bit @timmow @henrytk @combor @chrisfarms

### CODE REVIEW

File: cluster_spider.py
Comment (dcarley) [inline cluster_spider.py]: Can we pass the organisation name and cluster ID in as environment variables so that they're not hardcoded?

Partly for re-usability and partly because I'm not sure we should advertise our cluster ID - I don't think it's super sensitive, certainly not enough to scrub history, but we should avoid baking it in.

File: utils.py
Comment (dcarley) [inline utils.py]: We should document in the README that this is required. It might be good to call it something more agnostic of our setup, like `DEPLOYMENT_TAG`.

File: main.py
Comment (dcarley) [inline main.py]: I think we should check for the environment variables are startup and fail with an error if we don't have everything that we need. The error that you get at the moment is quite confusing:
```
➜  paas-compose-scraper git:(149806558-compose-metric-alerting) ✗ DEPLOY_ENV=foo pipenv run python3 main.py
OK
Traceback (most recent call last):
  File "main.py", line 9, in <module>
    utils.Report.ok()
  File "/Users/dcarley/projects/paas/paas-compose-scraper/utils.py", line 79, in ok
    report_health(CheckStatus.OK, message)
  File "/Users/dcarley/projects/paas/paas-compose-scraper/utils.py", line 69, in report_health
    check=check, host_name=host, status=status, message=message, tags=tags)
  File "/Users/dcarley/.local/share/virtualenvs/paas-compose-scraper-bR8x88GK/lib/python3.6/site-packages/datadog/api/service_checks.py", line 39, in check
    return super(ServiceCheck, cls)._trigger_action('POST', 'check_run', **params)
  File "/Users/dcarley/.local/share/virtualenvs/paas-compose-scraper-bR8x88GK/lib/python3.6/site-packages/datadog/api/resources.py", line 215, in _trigger_action
    return APIClient.submit(method, name, params)
  File "/Users/dcarley/.local/share/virtualenvs/paas-compose-scraper-bR8x88GK/lib/python3.6/site-packages/datadog/api/api_client.py", line 87, in submit
    raise ApiNotInitialized("API key is not set."
datadog.api.exceptions.ApiNotInitialized: API key is not set. Please run 'initialize' method first.
```


Can we only pass over the environment variables that we need, or is a full environment needed?

File: responses/fixtures/compose.html
Comment (dcarley) [inline responses/fixtures/compose.html]: Similar to a previous comment about variablising the org name and cluster ID, can we redact them from the fixture file?

File: main.py
Comment (dcarley) [inline main.py]: Opening a new subprocess feels a bit odd. I'm kind of surprised that the initialised DataDog library is available from the subprocess. Is it possible to use scrapy as a library from the same Python process?

File: settings.py
Comment (dcarley) [inline settings.py]: Thanks for making configurable 👍

**Review by dcarley — CHANGES_REQUESTED** [review]
Firstly well done 👏 

I tested this with #1031 - I know you said not to until this PR was merged, but it was easier to run that way. Few things that didn't fit into line-by-line review:

1. I've pushed a commit to run the tests automatically in Travis and have configured the relevant hooks. Can someone check that they're happy with it?
1. I've noticed that one host is being reported twice for each 10min interval. Anyone know why?

    screen shot 2017-09-14 at 15 18 42
1. We need to create a PR for https://github.com/alphagov/paas-credentials/compare/compose-billing
1. `requests` is logging in DEBUG mode, as seen from `cf logs paas-compose-scraper`, but I can't see it being explicitly configured. Do we need it? Should we disable it to make the logs easier to read and retain more history about the scraping?

**Comment by dcarley** [comment]
NB: I used `SELF_UPDATE_PIPELINE=false` when testing. I didn't re-run the paas-bootstrap pipelines with `ENABLE_DATADOG=true` because I didn't need the host metrics to test this functionality.

**Comment by 46bit** [comment]
I don't think I should review this, but I discussed self-updating pipelines with @alext earlier. `SELF_UPDATE_PIPELINE=false` is needed because `ENABLE_DATADOG` is not a [flag forwarded through](https://github.com/alphagov/paas-cf/blob/master/concourse/pipelines/create-cloudfoundry.yml#L341-L348) `self-update-pipelines` when it re-makes the pipeline config.

If we forwarded `ENABLE_DATADOG` then I believe a dev Concourse would keep Datadog enable until you explicitly. There's a few ways one can work with Concourse, from custom Makefiles to providing flags like we do here, and it's tricky to keep them all available. I think @alext is going to think a bit about this :-)

File: settings.py
Comment (46bit) [reply settings.py]: Candidate for optional environment variable?

File: main.py
Comment (46bit) [reply main.py]: Agreed, and the Datadog library is reinitialised inside the subprocess.

File: cluster_spider.py
Comment (chrisfarms) [reply cluster_spider.py]: 👍

File: utils.py
Comment (chrisfarms) [reply utils.py]: Don't feel strongly about renaming it tbh... but I think I'll change it to just allow setting a csv of tags from environment to leave it up to the user.

`DATADOG_TAGS="deployment:prod,whatever:treva"`

File: responses/fixtures/compose.html
Comment (chrisfarms) [reply responses/fixtures/compose.html]: Yep 🙌  ...did this when adding `COMPOSE_ACCOUNT_NAME` and `COMPOSE_CLUSTER_ID` env vars

File: settings.py
Comment (chrisfarms) [reply settings.py]: Yeah, why not... I'll make it configurable from environment

File: main.py
Comment (chrisfarms) [reply main.py]: The passing of the entire environment is mainly just an artifact of the `subprocess` method of executing the scraper.

However the shouty stacktrace errors have been addressed by other changes. (an env helper to gracefully bow out on missing required vars and making setting tags optional via `DATADOG_TAGS` envvar

Comment (chrisfarms) [reply main.py]: Heh, ok think I misunderstood ... willfix 👍

Comment (dcarley) [reply main.py]: I was thinking about whether we need to pass over all of `os.environ` because I was finding it difficult to understand, at a high level without reading further down the code, what environment variables were required to run. It felt like they could be function arguments instead. It sounds like the env helper will help though.

File: main.py
Comment (chrisfarms) [reply main.py]: Seem like it uses twisted under the hood ... so might not be simplest library import without the helper runner, but I'll have a poke around.

I think another issue with this loop is that Popen is async ... so as it stands it will run a new process every 5mins regardless of whether the last one complete ... which could be leak (although unlikely given the 5min timer).

That first `init_datadog` call is just to allow for the `Report.ok()` to phone in I believe ... not the main reporting.

Comment (46bit) [reply main.py]: I think it would be best if `utils.Report.ok()` was only called if we successfully got all the data. But at present there's no easy way to code that.

**Comment by chrisfarms** [comment]
Isn't the "ok" check somewhat redundant? ... The monitor could just alert
on "no data" for X periods to the same effect.... Unless I misunderstood
it's purpose?

**Review by dcarley — APPROVED** [review]

**Comment by dcarley** [comment]
Changes look good! Thanks @46bit and @chrisfarms !

**Pull request merged**
